### PR TITLE
Step 7 WI-1/2/3/4/5: Scratchpad IntelliSense — Shell API types, TS Plugin, CompletionItemProvider, Dynamic Schema, Refinements, Shared Logic

### PR DESCRIPTION
**Pull request by tnaum-ms**
## Goal

Add IntelliSense (autocompletion, hover docs, signature help) to DocumentDB scratchpad (`.documentdb.js`) files using a two-layer architecture.

This PR includes **Work Items 1, 2, 3, 4, and 5** of Step 7: Scratchpad CompletionItemProvider.

### WI-1: Foundation - `.d.ts` Type Definitions + TS Server Plugin ✅

- **Phase 1.1:** DocumentDB shell API `.d.ts` type definitions (865 lines, original JSDoc)
  - 10 BSON classes, 17 BSON constructors, 7 shell globals, EJSON utility
  - Database interface (14 methods), Collection interface (25 methods)
  - FindCursor (18 methods), AggregationCursor (8 methods)
  - Uses intersection type for `DocumentDBDatabase` to support both named methods and dynamic collection access (`db.anyName.`)

- **Phase 1.2:** TypeScript Server Plugin using **Inline Snapshot Injection** (Approach F)
  - Reads `.d.ts` content once at plugin load time
  - Proxies `getScriptSnapshot()` to prepend `.d.ts` content inline to scratchpad files
  - Proxies 20 position-based `LanguageService` methods with constant character-offset arithmetic
  - Works for virtual/untitled documents (unlike the initial `getExternalFiles()` approach which failed for untitled files)

- **Phase 1.3:** Language config, TS extension activation, TS server restart on first scratchpad open

### WI-2: Custom `CompletionItemProvider` - Core Engine ✅

- Two-stage cursor context detection (JS-level + inner query-object context)
- Dynamic collection names from `SchemaStore` after `db.` and in string literals
- DocumentDB API query operators (`$gt`, `$regex`, etc.) and field names inside method arguments
- 18 TDD behavior contracts for context detection
- Deduplication: Layer 2 only provides what Layer 1 (TS Plugin) cannot
- Shell API types package (`@vscode-documentdb/documentdb-shell-api-types`) with method registry and compatibility verification script

### WI-3: Dynamic Schema Integration + Collection Name Discovery ✅

- **Phase 3.1:** `CollectionNameCache` - reads collection names synchronously from `ClustersClient`'s in-memory cache (populated by tree view expansion/refresh). If no cached data exists, a single background `listCollections()` bootstraps the cache. No per-execution network calls — tree refresh keeps data current. Merges with SchemaStore for collections discovered through queries. 8 TDD behavior contracts.
- **Phase 3.2/3.3:** "Discover fields in collection..." action completion when SchemaStore has no data, plus `scanCollectionSchema` command (samples ~100 docs via `$sample`). Field detail uses `BSONTypes.toDisplayString()`.
- **ClustersClient caching:** In-memory cache for `listCollections`/`listDatabases` with synchronous accessors.
- **getCollection chain:** `db.getCollection('name').find({})` correctly resolves collection name for field completions and query operators.

### WI-4: Refinements and Edge Cases ✅

- **Phase 4.1: Sort prefix refinement** - All `sortText` values prefixed with `!` (ASCII 33) to sort above TS service completions (AbortController, Array, etc.). Fields use `!00_`, type-relevant operators `!0_`, comparison universals `!1a_`, BSON `!3_`, non-matching `!5_`.

- **Phase 4.3: ScratchpadHoverProvider** - New `vscode.HoverProvider` for `documentdb-scratchpad`:
  - Operator/BSON hover via `getAllCompletions()` with `$`-prefix fallback
  - Field hover via SchemaStore with `BSONTypes.toDisplayString()`, sparse indicator, multi-type display
  - Handles `$` cursor position (reads full operator name after `$`)
  - Handles quoted dotted field names via `extractQuotedKey` (reused from collection view)
  - Handles `db.getCollection("name").find({})` pattern (falls back to scanning from before string start)
  - Exported `getScratchpadHoverContent()` pure function for testability
  - 17 TDD behavior contracts

- **Phase 4.4: Edge cases** - `standalone === false` operator filtering (prevents nested-only operators like `$box`, `$geometry` from appearing in standard completions)

- **Type-aware value suggestions** - Reuses `getTypeSuggestionDefs()` from the collection view (`typeSuggestions.ts`):
  - Boolean: `true`, `false`
  - Numbers: range query `{ $gt: min, $lt: max }`, minimum value
  - Strings: regex pattern, string literal
  - Dates: ISODate, date range, last N days
  - ObjectId, null, arrays ($elemMatch, $size)

### WI-5: Shared Completion Logic Extraction + Refactor ✅

Structural refactoring — no user-visible changes. Extracts platform-neutral completion logic into a `shared/` module so both the scratchpad provider (VS Code API) and query editor provider (Monaco API) consume the same core.

- **Phase 5.1:** Extracted 7 platform-neutral modules into `src/webviews/documentdbQuery/shared/`:
  - Moved via `git mv` (preserves history): `completionKnowledge.ts`, `snippetUtils.ts`, `cursorContext.ts`, `extractQuotedKey.ts`
  - Extracted from mixed Monaco-specific modules: `sortPrefixes.ts`, `typeSuggestionData.ts`, `jsGlobalDefs.ts`
  - Consolidated `getVscodeOperatorSortPrefix()` duplicate — now wraps shared `getOperatorSortPrefix()`
  - Slimmed Monaco files: `typeSuggestions.ts` (269→67 lines), `jsGlobals.ts` (172→55 lines), `mapCompletionItems.ts` (158→110 lines)

- **Phase 5.2/5.3:** `CompletionData` interface + `CompletionKind` enum as platform-neutral contract in `shared/types.ts`

- **42 TDD behavior contracts** covering sort prefixes, key-position operators, snippet utils, type suggestions, JS globals

### Infrastructure

- Disabled webpack filesystem cache (both `webpack.config.ext.js` and `webpack.config.views.js`)
- Replaced em-dashes with colons/hyphens in user-visible strings
- Fixed bare "MongoDB" reference to "MongoDB API" per terminology guidelines

## Architecture: Two-Layer Autocompletion

| Layer | Mechanism | What It Provides |
|---|---|---|
| **Layer 1: TS Plugin** | Inline Snapshot Injection into TS language service | Shell globals, `db.` methods, collection methods, cursor chains, BSON types, hover docs, signature help, variable tracking |
| **Layer 2: Custom Provider** | `vscode.languages.registerCompletionItemProvider` | Dynamic collection names, DocumentDB API query operators (`$gt`, `$match`, etc.), field names from SchemaStore, collection names in strings, type-aware value suggestions, "Discover fields" action |

## How it works

1. A TS Server Plugin is registered via `package.json` `typescriptServerPlugins`
2. On first scratchpad open, the TS extension is activated and the TS server is restarted to load the plugin
3. The plugin prepends the `.d.ts` content inline to every scratchpad file's script snapshot
4. TypeScript sees the declarations as part of the file — hover, completions, signature help all work automatically
5. A custom `CompletionItemProvider` adds dynamic completions (collection names, query operators, field names, type suggestions) that TS cannot provide
6. A custom `HoverProvider` adds DocumentDB-specific hover docs (operators, BSON constructors, field types)
7. `CollectionNameCache` proactively fetches collection names using `ClustersClient` + in-memory caching
8. When no schema exists for a collection, a "Discover fields" action samples ~100 documents
9. Platform-neutral completion logic is shared between scratchpad and collection view query editors via `shared/` module

## Tests

- 1160 tests pass across 65 test suites
- 85 TDD behavior contracts:
  - 18 scratchpad context detection
  - 8 collection name cache / dynamic schema
  - 17 hover provider + edge cases
  - 42 shared completion logic (sort prefixes, operators, snippets, type suggestions, JS globals)
- Zero tsc compilation errors
- Zero ESLint errors

### CODE REVIEW

File: src/documentdb/ClustersExtension.ts [hunk 1]
```diff
@@ -270,6 +273,45 @@ export class ClustersExtension implements vscode.Disposable {
...
+                                // restarting it. Without this delay, the restart command can
+                                // arrive while the server is still starting, causing a crash.
+                                await new Promise((resolve) => setTimeout(resolve, 2000));
+                                // Restart the TS server so it picks up our plugin from
+                                // contributes.typescriptServerPlugins. Without this, the server
+                                // may have started before our extension was loaded and won't
+                                // have our plugin in --globalPlugins.
+                                await vscode.commands.executeCommand('typescript.restartTsServer');
+                            }
+                        }
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/ClustersExtension.ts]: Fixed in be31ee9. Extracted the async restart logic into `ensureTsRestart()` with try/catch error guard (logs to output channel). Also added a check for `vscode.workspace.textDocuments` during activation to handle already-open scratchpad files (hot-exit scenario).

File: src/documentdb/scratchpad/completions/scratchpadContextDetector.ts [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+        const argumentText = text.substring(argStart, offset);
+        const cursorOffset = offset - argStart;
+        return {
+            kind: 'method-argument',
+            methodName: chain.insideArgOf.methodName,
+            collectionName: chain.insideArgOf.collectionName,
+            argumentText,
+            cursorOffset,
+        };
+    }
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/scratchpad/completions/scratchpadContextDetector.ts:98]: Known issue, tracked in `docs/plan/future-pre-shell.md` item #4 (P2, ~1 day). The `insideArgOf` branch is vestigial scaffolding from an earlier iteration — the working path uses `detectMethodArgContext()` directly. Added code comments in cb332af pointing to the future work item.

File: src/documentdb/scratchpad/tsPlugin/index.ts [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+                    adjusted.push(start, result.spans[i + 1], result.spans[i + 2]);
+                }
+            }
+            return { spans: adjusted, endOfLineState: result.endOfLineState };
+        };
+
+        proxy.getEncodedSyntacticClassifications = (fileName: string, span: ts.TextSpan): ts.Classifications => {
+            const result = ls.getEncodedSyntacticClassifications(fileName, span);
+            if (!isScratchpadFile(fileName)) {
+                return result;
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/scratchpad/tsPlugin/index.ts:361]: Fixed in 2e14f21. Added input span shift by `prefixLength` for both `getEncodedSemanticClassifications` and `getEncodedSyntacticClassifications`, consistent with how all 15 other proxy methods use `adjustPosition()`.

File: src/documentdb/scratchpad/typeDefs/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Type Definitions — Moved
+
+The `documentdb-shell-api.d.ts` file has been moved to the
+`@vscode-documentdb/shell-api-types` package at:
+
+    packages/shell-api-types/typeDefs/documentdb-shell-api.d.ts
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/scratchpad/typeDefs/README.md]: Fixed in f013906. Updated all 3 references in this file from `shell-api-types` to `documentdb-shell-api-types`.

File: packages/documentdb-shell-api-types/scripts/verify-compatibility.ts [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+ *   1. Fetches the compatibility page from the azure-databases-docs GitHub repo
+ *   2. Extracts the "Database commands" table to find supported commands
+ *   3. Checks that every server command in our method registry is still
+ *      marked as supported (✅) in the official documentation
+ *   4. Reports any mismatches (commands we reference that are no longer supported,
+ *      or new commands that we might want to add)
+ *
+ * Usage:
+ *   npm run verify
+ *   npx ts-node packages/shell-api-types/scripts/verify-compatibility.ts
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply packages/documentdb-shell-api-types/scripts/verify-compatibility.ts]: Fixed in f013906. Updated usage example path from `packages/shell-api-types/` to `packages/documentdb-shell-api-types/`.

File: packages/documentdb-shell-api-types/package.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@vscode-documentdb/documentdb-shell-api-types",
+  "version": "1.0.0",
+  "description": "Shell API type definitions and method-to-command mapping for DocumentDB scratchpad IntelliSense",
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "files": [
+    "dist"
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply packages/documentdb-shell-api-types/package.json:8]: The `files` field concern is valid but premature — this package is consumed only within the monorepo via webpack, not published to npm. Added a future work item (nice-to-have #12: Pre-npm Publishing) to track the packaging fixes needed before standalone publication.

File: packages/documentdb-shell-api-types/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+/**
+ * @vscode-documentdb/shell-api-types
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply packages/documentdb-shell-api-types/src/index.ts]: Fixed in f013906. Updated banner comment from `@vscode-documentdb/shell-api-types` to `@vscode-documentdb/documentdb-shell-api-types`.

File: src/documentdb/scratchpad/completions/ScratchpadHoverProvider.ts [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+            if (match.description) {
+                lines.push(match.description);
+            }
+            if (match.link) {
+                lines.push(`[ⓘ Documentation](${match.link})`);
+            }
+
+            return {
+                contents: [{ value: lines.join('\n\n'), isTrusted: true, supportHtml: true }],
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/documentdb/scratchpad/completions/ScratchpadHoverProvider.ts:76]: Keeping `isTrusted: true` and `supportHtml: true`. The hover content uses `<br>` HTML tags for spacing (requires `supportHtml`) and renders `[ⓘ Documentation](url)` links to external docs (requires `isTrusted` for reliable link rendering across all VS Code hover contexts). The operator data is static and first-party, so the attack surface is minimal.
